A15-binary-logging.md

```diff
@@ -0,0 +1,375 @@
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    ClientHeader client_header = 6;
+    ServerHeader server_header = 7;
+    // Used by EVENT_TYPE_CLIENT_MESSAGE, EVENT_TYPE_SERVER_MESSAGE
+    Message message = 8;
+    Trailer trailer = 9;
+  }
+
+  // true if payload does not represent the full message or metadata.
+  bool payload_truncated = 10;
+
+  // Peer address information, will only be recorded in
+  // EVENT_TYPE_SERVER_HEADER for client side or
+  // EVENT_TYPE_CLIENT_HEADER for server side.
+  // On client side, the remote peer is not yet known when the header
+  // is sent because load balancing has not yet happened.
+  // On server side, it is possible that the RPC is cancelled before
+  // it sends the header.
+  Peer peer = 11;
+};
+
+message ClientHeader {
+  // This contains only the metadata from the application.
+  Metadata metadata = 1;
+
+  string method_name = 2;
+
+  string authority = 3;
+
+  // the RPC timeout
+  google.protobuf.Duration timeout = 4;
+
+}
+
+message ServerHeader {
+  // This contains only the metadata from the application.
+  Metadata metadata = 3;
+}
+
+message Trailer {
+  // status_code and status_message:
+  // Only present for EVENT_TYPE_SERVER_TRAILER.
+  uint32 status_code = 1;
+
+  // An original status message before any transport specific
+  // encoding.
+  string status_message = 2;
+
+  // The value of the 'grpc-status-details-bin' metadata key. If
+  // present, this is always an encoded 'google.rpc.Status' message.
+  bytes status_details = 3;
+}
+
+// Message payload, used by CLIENT_MESSAGE and SERVER_MESSAGE
+message Message {
+  // Length of the message. It may not be the same as the length of the
+  // data field, as the logging payload can be truncated or omitted.
+  uint32 length = 1;
+  // May be truncated or omitted.
+  bytes data = 2;
+}
+
+// A list of metadata pairs, used in the payload of client header,
+// server header and trailer.
+// Implementations may omit some entries to honor the header limits
+// of GRPC_BINARY_LOG_CONFIG.
+//
+// Header keys added by gRPC are omitted. To be more specific,
+// implementations will not log the following entries, and this is
+// not to be treated as a truncation:
+// - entries handled by grpc that are not user visible, such as those
+//   that begin with 'grpc-' (with exception of grpc-trace-bin)
+//   or keys like 'lb-token'
+// - transport specific entries, including but not limited to:
+//   ':path', ':authority', 'content-encoding', 'user-agent', 'te', etc
+// - entries added for call credentials
+//
+// Implementations must always log grpc-trace-bin if it is present.
+// The pair will count towards the size limit,
+message Metadata {
+  repeated MetadataEntry entry = 1;
+}
+
+// A metadata key value pair
+message MetadataEntry {
+  bytes key = 1;
+  bytes value = 2;
+}
+
+// Peer information
+message Peer {
+  enum PeerType {
```

Comment:
Nit/optional: this could be simply "Type" since it's nested inside Peer.

Reply:
Done